utils/singlesimulation.py

```diff
@@ -153,16 +158,188 @@ def run(self, ranker, output_key):
 
     run_results = []
     impressions = 0
+
+    text = open("Weights.txt","r")
```

Comment:
We should have a separate file for attack instead of making such a big change to the original singlesimulation.py . This makes it hard to read and run the no attacker setting.

You should also include the script / command line with parameters / readme file to run the code.